core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1337,13 +1337,16 @@ class UnifiedLog(@volatile var logStartOffset: Long,
       } else if (targetTimestamp == ListOffsetsRequest.MAX_TIMESTAMP) {
         // Cache to avoid race conditions. `toBuffer` is faster than most alternatives and provides
         // constant time access while being safe to use with concurrent collections unlike `toArray`.
-        val segmentsCopy = logSegments.asScala.toBuffer
-        val latestTimestampSegment = segmentsCopy.maxBy(_.maxTimestampSoFar)
-        val latestTimestampAndOffset = latestTimestampSegment.readMaxTimestampAndOffsetSoFar
-
-        Some(new TimestampAndOffset(latestTimestampAndOffset.timestamp,
-          latestTimestampAndOffset.offset,
-          latestEpochAsOptional(leaderEpochCache)))
+        val latestTimestampSegment = logSegments.asScala.toBuffer.maxBy[Long](_.maxTimestampSoFar)
+        // cache the timestamp and offset
+        val maxTimestampSoFar = latestTimestampSegment.readMaxTimestampAndOffsetSoFar
+        // lookup the position of batch to avoid extra I/O
+        val position = latestTimestampSegment.offsetIndex.lookup(maxTimestampSoFar.offset)
+        val lpc = latestEpochAsOptional(leaderEpochCache)
+        Some(latestTimestampSegment.log.batchesFrom(position.position).asScala
+          .find(_.maxTimestamp() == maxTimestampSoFar.timestamp)
+          .flatMap(batch => batch.offsetOfMaxTimestamp().asScala.map(new TimestampAndOffset(batch.maxTimestamp(), _, lpc)))
+          .getOrElse(new TimestampAndOffset(-1, 0, lpc))) // always return something for backward compatibility
```

Comment:
@showuon your previous comment is right (sorry that I can't find the comment but it is in my mind). We need to return `offset=0 and ts=-1` if there are no batches for the sake of backward compatibility.

Reply:
Hmm, why do we need to return offset=0 when we can't find the maxTimestamp?

Reply:
> why do we need to return offset=0 when we can't find the maxTimestamp?

oh, my bad. we should return base offset instead of zero.